docs/source/cpp/dataset.rst

```diff
@@ -0,0 +1,381 @@
+   :language: cpp
+   :lines: 87-104
+   :linenos:
+   :lineno-match:
+
+then we can read the Feather file using the same functions, but passing a
+:class:`arrow::dataset::IpcFileFormat`:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 310,323
+   :linenos:
+
+Customizing file formats
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+:class:`arrow::dataset::FileFormat` objects have properties that control how
+they are read. For example::
+
+  auto format = std::make_shared<ds::ParquetFileFormat>();
+  format->reader_options.dict_columns.insert("a");
+
+Will configure column ``"a"`` to be dictionary-encoded when read.
+
+.. _cpp-dataset-filtering-data:
+
+Filtering data
+--------------
+
+So far, we've read the entire dataset, but this can be inefficient or use too
+much memory. The :class:`arrow::dataset::Scanner` offers control over what data
+to read.
+
+:func:`arrow::dataset::ScannerBuilder::Project` lets us read only specified
+columns:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 172-191
+   :emphasize-lines: 16
+   :linenos:
+   :lineno-match:
+
+For formats which support pushdown (such as Parquet), this means that only the
+specified columns will be read from the filesystem, saving I/O costs.
+
+A filter can be provided with :func:`arrow::dataset::ScannerBuilder::Filter`,
+so that rows which do not match the filter predicate will not be included in
+the returned table.
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 172-191
+   :emphasize-lines: 17
+   :linenos:
+   :lineno-match:
+
+.. TODO Expressions not documented pending renamespacing
+
+Projecting columns
+------------------
+
+In addition to selecting columns,
+:func:`arrow::dataset::ScannerBuilder::Project` can also be used for more
+complex projections in combination with expressions.
+
+In this case, we pass a vector of expressions used to construct column values,
+and a vector of names for the columns:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 193-223
+   :emphasize-lines: 18-28
+   :linenos:
+   :lineno-match:
+
+This also determines the column selection; only the given columns will be
+present in the resulting table. If you want to include a derived column in
+*addition* to the existing columns, you can build up the expressions from the
+dataset schema:
+
+.. literalinclude:: ../../../cpp/examples/arrow/dataset_documentation_example.cc
+   :language: cpp
+   :lines: 225-254
+   :emphasize-lines: 17-27
+   :linenos:
+   :lineno-match:
+
+Reading and writing partitioned data
+------------------------------------
+
+So far, we've been working with datasets consisting of flat directories with
+files. However, a dataset can exploit nested directory structures defining a
```

Comment:
The whole sentence starting with `However` is a little awkward to me.  Maybe...

```
However, a "partitioned" dataset relies on nested directory structures where sub-directory names hold describe the data stored in that directory
```

...but I'm not really sold on that either.